### PR TITLE
Tweak paginator to accommodate {slug}.html patterns

### PR DESCRIPTION
**Pull request by oulenz**
Fixes https://github.com/getpelican/pelican/issues/1137, fixes https://github.com/getpelican/pelican/issues/1615, fixes https://github.com/getpelican/pelican/issues/2155, fixes .
Makes pr https://github.com/getpelican/pelican/pull/1956 redundant.

At present, pagination produces critical errors with the default `*_SAVE_AS` settings for authors, categories and tags, because the `{base_name}` variable strips `{slug}.html` from `foo/{slug}.html`. This fixes that by ensuring that `{base_name}` only strips `.html` and `/index.html`

### CODE REVIEW

**Comment by justinmayer** [comment]
Would any @getpelican/reviewers be so kind as to review this pull request and either approve it or provide feedback/comments?

File: docs/settings.rst [hunk 1]
```diff
@@ -1000,33 +1000,39 @@ You can use the following settings to configure the pagination.
...
+By default, pages subsequent to ``.../foo.html`` are created as
+``.../foo2.html``, etc. The ``PAGINATION_PATTERNS`` setting can be used to
+change this. It takes a sequence of triples, where each triple consists of::
 
   (minimum page, URL setting, SAVE_AS setting,)
 
-For example, if you wanted the first page to just be ``/``, and the
-second (and subsequent) pages to be ``/page/2/``, you would set
-``PAGINATION_PATTERNS`` as follows::
+For ``URL setting`` and ``SAVE_AS setting``, you may use a number of variables.
```
Comment (avaris) [inline docs/settings.rst]:
```
For ``URL`` setting and ``SAVE_AS`` setting ...
```

File: pelican/paginator.py [hunk 2]
```diff
@@ -67,8 +67,11 @@ def _get_page_range(self):
 
 class Page(object):
     def __init__(self, name, url, object_list, number, paginator, settings):
+        self.full_name = name
         self.name, self.extension = os.path.splitext(name)
-        self.base_url = url
+        dirname, filename = os.path.split(name)
+        self.base_name = dirname if filename == 'index.html' else self.name
+        self.full_url = url
```
Comment (avaris) [inline pelican/paginator.py]: Someone might be using `base_url` in a template. No need to break this with no reason.

File: pelican/paginator.py [hunk 3]
```diff
@@ -67,8 +67,11 @@ def _get_page_range(self):
 
 class Page(object):
     def __init__(self, name, url, object_list, number, paginator, settings):
+        self.full_name = name
         self.name, self.extension = os.path.splitext(name)
-        self.base_url = url
+        dirname, filename = os.path.split(name)
+        self.base_name = dirname if filename == 'index.html' else self.name
```
Comment (avaris) [inline pelican/paginator.py]: `index.htm` 🤔

File: pelican/paginator.py (hunk 2 above)
Comment (oulenz) [reply pelican/paginator.py]: If I parse this method (`_from_settings`) correctly, it's the values in `context` that get inserted into the patterns. I renamed the attribute from `base_url` to `full_url`, but I didn't rename the corresponding value in `context`, which is still `url`. So this shouldn't break anything, right?

File: docs/settings.rst (hunk 1 above)
Comment (oulenz) [reply docs/settings.rst]: Thanks, renamed those variables

File: pelican/paginator.py (hunk 3 above)
Comment (oulenz) [reply pelican/paginator.py]: Thanks, added `index.htm`

File: pelican/paginator.py (hunk 2 above)
Comment (avaris) [reply pelican/paginator.py]: `Page` object gets passed to the template. Someone might be using `articles_page.base_url` in their templates... for something...

Comment (oulenz) [reply pelican/paginator.py]: Ah ok! I think the `base_` part is misleading, but I guess we're stuck with it. I've reverted to `base_url`.

Comment (avaris) [reply pelican/paginator.py]: Joys of backward compatibility :)

**Comment by justinmayer** [comment]
Sincere thanks to @oulenz for the fix and to @avaris for reviewing! 😁

**Pull request merged**